File: src/services/tauri/sshShell.ts
import { invoke } from "@tauri-apps/api/core";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";

export interface SshTerminalDataPayload {
  shellId: string;
  data: string;
}

export interface SshTerminalClosedPayload {
  shellId: string;
}

export async function openSshShell(cols: number, rows: number): Promise<string> {
  return invoke<string>("open_ssh_shell", { cols, rows });
}

export async function writeSshShell(
  shellId: string,
  data: string,
): Promise<void> {
  await invoke<void>("write_ssh_shell", { shellId, data });
}

export async function resizeSshShell(
  shellId: string,
  cols: number,
  rows: number,
): Promise<void> {
  await invoke<void>("resize_ssh_shell", { shellId, cols, rows });
}

export async function closeSshShell(shellId: string): Promise<void> {
  await invoke<void>("close_ssh_shell", { shellId });
}

export async function closeAllSshShells(): Promise<void> {
  await invoke<void>("close_all_ssh_shells");
}

export function listenSshTerminalData(
  handler: (payload: SshTerminalDataPayload) => void,
): Promise<UnlistenFn> {
  return listen<SshTerminalDataPayload>("ssh-terminal-data", (event) => {
    handler(event.payload);
  });
}

export function listenSshTerminalClosed(
  handler: (payload: SshTerminalClosedPayload) => void,
): Promise<UnlistenFn> {
  return listen<SshTerminalClosedPayload>("ssh-terminal-closed", (event) => {
    handler(event.payload);
  });
}

File: src-tauri/src/commands/ssh_shell.rs
use tauri::{AppHandle, State};

use crate::services::SshConnectionManager;

#[tauri::command]
pub fn open_ssh_shell(
    cols: u32,
    rows: u32,
    app: AppHandle,
    state: State<'_, SshConnectionManager>,
) -> Result<String, String> {
    state.open_shell(cols, rows, &app)
}

#[tauri::command]
pub fn write_ssh_shell(
    shell_id: String,
    data: String,
    state: State<'_, SshConnectionManager>,
) -> Result<(), String> {
    state.write_shell(&shell_id, &data)
}

#[tauri::command]
pub fn resize_ssh_shell(
    shell_id: String,
    cols: u32,
    rows: u32,
    state: State<'_, SshConnectionManager>,
) -> Result<(), String> {
    state.resize_shell(&shell_id, cols, rows)
}

#[tauri::command]
pub fn close_ssh_shell(
    shell_id: String,
    state: State<'_, SshConnectionManager>,
) -> Result<(), String> {
    state.close_shell(&shell_id)
}

#[tauri::command]
pub fn close_all_ssh_shells(state: State<'_, SshConnectionManager>) -> Result<(), String> {
    state.close_all_shells()
}
